### PR TITLE
Adds `BlockNumberProvider` in multisig, proxy and nft pallets

### PR DESCRIPTION
**Pull request by gupnik**
Step in https://github.com/paritytech/polkadot-sdk/issues/3268

This PR adds the ability for these pallets to specify their source of the block number. This is useful when these pallets are migrated from the relay chain to a parachain and vice versa.

This change is backwards compatible:
1. If the `BlockNumberProvider` continues to use the system pallet's block number
2. When a pallet deployed on the relay chain is moved to a parachain, but still uses the relay chain's block number

However, we would need migrations if the deployed pallets are upgraded on an existing parachain, and the `BlockNumberProvider` uses the relay chain block number.

### CODE REVIEW

File: substrate/frame/proxy/src/lib.rs [hunk 1]
```diff
@@ -392,7 +397,7 @@ pub mod pallet {
 			let announcement = Announcement {
 				real: real.clone(),
 				call_hash,
-				height: system::Pallet::<T>::block_number(),
+				height: T::BlockNumberProvider::current_block_number(),
```
Comment (franciscoaguirre) [inline substrate/frame/proxy/src/lib.rs:400]: The migration would need to convert all old block numbers into new ones. We could provide a migration for teams upgrading these pallets that assumes the old times were 12 seconds and the new ones 6 seconds since it's the relay chain?

Comment (joepetrowski) [reply substrate/frame/proxy/src/lib.rs:400]: Couldn't you just define three `const`s in the migration type (e.g. [here](https://github.com/paritytech/polkadot-sdk/blob/e542796/substrate/frame/identity/src/migration.rs#L70))? They would be `old_block_time` (12s), `new_block_time` (6s), and `offset` (to account for the different block heights).

Comment (Ank4n) [reply substrate/frame/proxy/src/lib.rs:400]: https://github.com/paritytech/polkadot-sdk/pull/5656#issuecomment-2352671254
^^ This would be a better strategy than migration.

File: substrate/frame/multisig/src/lib.rs [hunk 2]
```diff
@@ -168,6 +168,9 @@ pub mod pallet {
 
 		/// Weight information for extrinsics in this pallet.
 		type WeightInfo: WeightInfo;
+
+		/// Provider for the block number. Normally this is the `frame_system` pallet.
```
Comment (kianenigma) [inline substrate/frame/multisig/src/lib.rs:172]: It is quite unfortunate that we cannot add this to `parachain_system` and be done with it. The underlying issue is that all of our pallets rely on `frame_system`, not `parachain_system`. In other words, I pallets are written mostly with the intention to be used as a solochain, which in retrospect is not how it should have been.

TLDR, adding the type to all pallets is as good as we can do this. 

wdyt about marking `frame_system::block_number` as deprecated and explicitly asking everyone to decide on their block number source?

File: prdoc/pr_5723.prdoc [hunk 3]
```diff
@@ -0,0 +1,15 @@
+title: Adds `BlockNumberProvider` in multisig, proxy and nft pallets
+
+doc:
+  - audience: Runtime Dev
+    description: |
+      This would allow these pallets to use the relay chain block number when deployed on a parachain.
```
Comment (kianenigma) [inline prdoc/pr_5723.prdoc:6]: What this PR in fact is doing is adding the ability for pallets to specify their source of block number.

Then, you should describe in which case this change is deployed backwards compatible: RC -> RC, or Para -> Para. 

Then, you should describe if otherwise (RC -> Para, Para -> RC), what migratiion should be used. 

@davidk-pt is trying to introduce such a migration in https://github.com/paritytech/polkadot-sdk/pull/5656, I suggest you two to work together, first polish the `type BlockNumberProvider`, then work on a reusable migration, then move individual pallets.

Comment (kianenigma) [reply prdoc/pr_5723.prdoc:6]: Still is a relevant point.

**Review by kianenigma — APPROVED** [review]
LGTM as abstractions to add, excluding the possible need for a migration.

File: substrate/frame/multisig/src/lib.rs (hunk 2 above)
Comment (kianenigma) [reply substrate/frame/multisig/src/lib.rs:172]: > wdyt about marking frame_system::block_number as deprecated and explicitly asking everyone to decide on their block number source?

If this is something that you agree to, please make an issue for it.

Comment (gupnik) [reply substrate/frame/multisig/src/lib.rs:172]: > > wdyt about marking frame_system::block_number as deprecated and explicitly asking everyone to decide on their block number source?
> 
> If this is something that you agree to, please make an issue for it.

Done: https://github.com/paritytech/polkadot-sdk/issues/6245

**Comment by kianenigma** [comment]
The metric through which you have chosen to migrate Nft pallet seems wrong btw -- this pallet is not priority as it is not on the relay chain.

File: substrate/frame/multisig/src/lib.rs [hunk 4]
```diff
@@ -50,23 +50,9 @@ pub mod weights;
...
-};
-use frame_system::{self as system, pallet_prelude::BlockNumberFor, RawOrigin};
-use scale_info::TypeInfo;
-use sp_io::hashing::blake2_256;
-use sp_runtime::{
-	traits::{BlockNumberProvider, Dispatchable, TrailingZeroInput, Zero},
-	DispatchError, RuntimeDebug,
+use frame::{
+	prelude::*,
+	traits::{BlockNumberProvider, Currency, ReservableCurrency},
```
Comment (kianenigma) [inline substrate/frame/multisig/src/lib.rs:55]: `BlockNumberProvider` should go into the prelude, as it is quite useful to almost all pallets?

File: substrate/frame/proxy/src/lib.rs [hunk 5]
```diff
@@ -163,6 +163,9 @@ pub mod pallet {
 		/// into a pre-existing storage value.
 		#[pallet::constant]
 		type AnnouncementDepositFactor: Get<BalanceOf<Self>>;
+
+		/// Provider for the block number. Normally this is the `frame_system` pallet.
+		type BlockNumberProvider: BlockNumberProvider<BlockNumber = BlockNumberFor<Self>>;
```
Comment (gui1117) [inline substrate/frame/proxy/src/lib.rs:168]: Why do we force the `BlockNumber` to be the frame system block number type?

Instead we should allow block number to be any type here and redefine a helper type:

```rust
type BlockNumberFor<T> = <T as Config>::BlockNumberProvider as BlockNumberProvider>::BlockNumber;
```
```suggestion
		type BlockNumberProvider: BlockNumberProvider;
```

Comment (muharem) [reply substrate/frame/proxy/src/lib.rs:168]: I agree, we should not force the same type. But probably we cannot change the `BlockNumberFor` type reference now, since not all configs have the `BlockNumberProvider`

**Review by gui1117 — APPROVED** [review]
Looks good to me,
Using different types make the change easier to check.

**Review by pgherveou — APPROVED** [review]

**Review by xermicus — APPROVED** [review]
